### ISSUE TITLE
Null Reference Exceptions in EventSource

### ISSUE THREAD

**Issue by JeffAshton** [launchdarkly/dotnet-eventsource#24]
**Message:** Object reference not set to an instance of an object.
**Type:** System.NullReferenceException

```
   at LaunchDarkly.EventSource.EventSource.ProcessResponseContent(String content)
   at LaunchDarkly.EventSource.EventSourceService.<ProcessResponseFromReaderAsync>d__13.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LaunchDarkly.EventSource.EventSourceService.<ConnectToEventSourceApi>d__11.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LaunchDarkly.EventSource.EventSourceService.<GetDataAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LaunchDarkly.EventSource.EventSource.<ConnectToEventSourceAsync>d__37.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at LaunchDarkly.EventSource.EventSource.<StartAsync>d__32.MoveNext()
```
Occurs around the same time tasks are cancelled:

![image](https://user-images.githubusercontent.com/12101120/36539283-4a506da2-17a4-11e8-9f00-3cf7d171a57f.png)

**Comment by eli-darkly** [comment]
I'm awfully sorry we didn't follow up on this issue sooner - it seems to have completely fallen through the cracks somehow. We should have a fix pretty soon. I wanted to confirm, though: am I right that the exception did not propagate to application code, but was caught and logged inside the EventSource?

**Comment by eli-darkly** [comment]
Sorry to hear that. Hmm... I would not have expected this particular bug to cause updates to be lost; my understanding was that the exception would be caught in `StartAsync()`, and it would then reconnect as it does for any stream failure, at which point the complete state of things would be re-sent in the first message. But I may be wrong. I'll try causing an error deliberately at that point to confirm.

I presume you haven't noticed any other suspicious log activity preceding the times when updates stop happening...?

**Comment by JeffAshton** [comment]
Came across this one in the Stream Manager

[Null Reference in Stream Manager.pdf](https://github.com/launchdarkly/dotnet-eventsource/files/2421394/Null.Reference.in.Stream.Manager.pdf)

**Comment by JeffAshton** [comment]
Looks like you aren't handling your task cancellations gracefully

[Poor Handling of Task Cancelled Exceptions.pdf](https://github.com/launchdarkly/dotnet-eventsource/files/2421410/Poor.Handling.of.Task.Cancelled.Exceptions.pdf)

A pattern I find myself doing is something like this:

```
CancellationToken cancellationToken;

try {
   // something async with cancellation token

} catch( OperationCanceledException err ) when( err.CancellationToken.Equals( cancellationToken ) ) {
  // not an error

} catch( Exception err ) {
  // log the error
}
```

**Comment by eli-darkly** [comment]
@JeffAshton, on your last comment, I see what you're getting at and that extra `catch` seems fine, but I'm surprisingly unable to reproduce this behavior. I'm opening a stream, waiting a bit, then calling `Close()` on it which should trigger the cancellation token. Nothing happens except that the stream closes - my `OnError` is not called and no exception is logged. What am I missing?

**Comment by eli-darkly** [comment]
We may still need to fix the cancellation token issue, I just wasn't able to reproduce it so far. In the meantime, I've released v3.2.0 of LaunchDarkly.EventSource which should fix the null reference exception described here.

**Comment by JeffAshton** [comment]
@eli-darkly  Thanks.  I'll monitor and let you know
